File: 1114-print-in-order/1114-print-in-order.cpp
class Foo {
public:
    mutex mtx;
    condition_variable cv;
    int cur = 1;
    Foo() { }

    void first(function<void()> printFirst) {
        printFirst();
        cur = 2;
        cv.notify_all();
    }

    void second(function<void()> printSecond) {
        unique_lock<mutex> lock(mtx);
        cv.wait(lock, [this]{ return cur == 2; });
        printSecond();
        cur = 3;
        cv.notify_all();
    }

    void third(function<void()> printThird) {
        unique_lock<mutex> lock(mtx);
        cv.wait(lock, [this]{ return cur == 3; });
        printThird();
    }
};

File: 1114-print-in-order/1114-print-in-order.java
class Foo {
    private Semaphore sem2;
    private Semaphore sem3;
    
    public Foo() { 
        sem2 = new Semaphore(0);
        sem3 = new Semaphore(0);
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printFirst.run();
        sem2.release();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        sem2.acquire();
        printSecond.run();
        sem3.release();
    }

    public void third(Runnable printThird) throws InterruptedException {
        sem3.acquire();
        printThird.run();
    }
}